### PR TITLE
Clear unpublished services in termination

### PR DESCRIPTION
**Pull request by kletkavrubashku**

### CODE REVIEW

**Comment by kletkavrubashku** [comment]
I understood that fix can be better. WIP

File: src/context.cpp [hunk 1]
```diff
@@ -389,8 +392,8 @@ class context_impl_t : public context_t {
         m_config->services().each([&](const std::string& name, const config_t::component_t&){
             try {
                 actors.push_back(remove(name));
-            } catch (...) {
-                // A service might be absent because it has failed to start during the bootstrap.
+            } catch (const std::exception& e) {
+                COCAINE_LOG_WARNING(m_log, "termination of service '{}' is failed: {}", name, e);
```
Comment (shaitan) [inline src/context.cpp]:
```suggestion
                COCAINE_LOG_WARNING(m_log, "termination of service '{}' has failed: {}", name, e);
```

File: src/context.cpp [hunk 2]
```diff
@@ -309,19 +309,22 @@ class context_impl_t : public context_t {
...
+            auto service = std::move(it->second);
+            list.erase(it);
+            return service;
+        };
 
-        service->terminate();
+        auto service = m_services.apply(extract_service);
+        if (service) {
+            service->terminate();
+        } else if (!(service = extract_service(m_unpublished))) {
```
Comment (shaitan) [inline src/context.cpp:326]: Unpublished service should be terminated too, shouldn't it?

Comment (kletkavrubashku) [reply src/context.cpp:326]: Yeah, I was gonna do that.
https://github.com/interiorem/cocaine-core/pull/6#issuecomment-437189240

Termination is not required for unpublished actor and lead to crash, but it fixable and will be better for using. I agree.

Comment (kletkavrubashku) [reply src/context.cpp:326]: We decided to do it in separate task. I added TODO.

**Review by shaitan — APPROVED** [review]

**Pull request merged**